File: Computational_Physics_2/1.Lectures/8.Computo_en_paralelo_OPENMP/examples/example_3_Matrix_Multiplication/C/MatrMult_ser.c
#include <stdio.h>
#include <stdlib.h>
#ifdef _OPENMP
#include <omp.h>
#endif

int main(  int argc, char *argv[] )
{ 

   double *A, *B, *C, *tC;
   int L, M, N, L0, L1;
   double cput2, cput1, wall2, wall1;

   int i, j, k, t, st;
   double v, rmin, rmax;

/*  OMP variables */
   int ierr, my_rank, numprocs;

      fprintf(stdout,"Input dimensions of matrices  L, M, N:\n");
      fscanf(stdin,"%d, %d, %d",&L, &M, &N);
      fprintf(stdout," Matrices dimensions are: %d, %d, %d\n",L,M,N);

         A = (double*) malloc(sizeof((double)1.0)*L*M);
         B = (double*) malloc(sizeof((double)1.0)*N*M);
         C = (double*) malloc(sizeof((double)1.0)*L*N);
         if ( A == NULL || B == NULL || C == NULL ) {
            fprintf(stderr,"Error allocating matrices \n");
      	    return(-1);
         }


   /*  Initializations */
   // Insert parallel region, pay attention to  shared or private variables
   

   	// Do it in parallel 	
   	   for ( i = 0; i < L*M; i++ ) A[i] = 0.0;
   	
   	
   	// Do it in parallel
   	   for ( i = 0; i < N*M; i++ ) B[i] = 0.0;
   	  
      
   	// Do it in parallel
   	   for ( i = 0; i < L*N; i++ ) C[i] = 0.0;
   	  

   	// Do it in parallel
   	   for ( i = 0; i < L; i++ ) {
   	   	   for ( k = 0; k < M; k++ ) {
   	   	   	   A[ i*M + k] = (double)(i-k) / (double)(i+k+1);
   	   	   }
   	   }
   
   	// Do it in parallel
   	  for ( j = 0; j < N; j++ ) {
   	  	  for ( k = 0; k < M; k++ ) {
   	  	  	  B[ k*N + j ] = 0.01 * (double)(k-j) / (double)(k+j+1);
   	  	  }
   	  }
   // Close parallel region
   // This statement is executed only by master thread if openmp is defined
   //wall1 = omp_get_wtime();
   

   /* Compute local */
   
   // Insert parallel region, pay attention to  shared or private variables
   
   // This statement is executed only if openmp is defined
   //my_rank = omp_get_thread_num();
   //numprocs = omp_get_num_threads();
   
   my_rank = 0; numprocs = 1;
   
   ierr = intervals(L,&L0,&L1,numprocs,my_rank);
   
   /*  Compute multiplication */
   // This statement is executed by all threads but only one at a time
      printf("Process %d of %d computes C(:,%d:%d) \n",my_rank,numprocs,L0,L1-1);
      
      for ( j = 0; j < N; j++ ) {
         for ( i = L0; i < L1; i++ ) {
            C[ i*N + j ] = 0.0;
            for ( k = 0; k < M; k++ ) {
               C[ i*N + j ] = C[ i*N + j ] + A[ i*M + k ] * B[ k*N + j ];
            }
         }
      }
   // Close parallel region
   
   // This statement is executed only by master thread if openmp is defined
     // wall2 = omp_get_wtime();
   
      /* Compare results */ 
      tC = (double*) malloc(sizeof((double)1.0)*L*N);
      if ( tC == NULL ) {
         fprintf(stderr," Process %d: Error allocating matrix tC \n",my_rank);
      }

      for ( i = 0; i < L; i++ ) {
         for ( k = 0; k < M; k++ ) {
            A[ i*M + k] = (double)(i-k) / (double)(i+k+1);
         }
      }
   
      for ( j = 0; j < N; j++ ) {
         for ( i = 0; i < L; i++ ) {
            tC[ i*N + j ] = 0.0;
            for ( k = 0; k < M; k++ ) {
               tC[ i*N + j ] = tC[ i*N + j ] + A[ i*M + k ] * B[ k*N + j ];
            }
         }
      }

      rmin = rmax = tC[0] - C[0];
      for ( j = 0; j < N; j++ ) {
         for ( i = 0; i < L; i++ ) {
            v = tC[ i*N + j ] - C[ i*N + j ];
            if ( v < rmin ) {
               rmin = v;
            }
            if ( v > rmax ) {
               rmax = v;
            }
         }
      }

      fprintf(stderr," Check results: rmin, rmax = %f, %f\n",rmin,rmax);
      
      fprintf(stdout,"Elapsed time is %lf \n",(wall2-wall1));
      	   	   
  
   return(0);
}

int intervals(int length, int* start, int* end, int segments, int n)
{
    int delta;

    delta = ( length + (segments - 1) ) / segments;
    *start = delta * n; *end = delta * ( n + 1);
    if ( *end > length ) *end = length;
   
   return(0);
}

File: Computational_Physics_2/1.Lectures/8.Computo_en_paralelo_OPENMP/examples/example_3_Matrix_Multiplication/Fortran/MatrMult_ser.f90
PROGRAM MatrMult 
   IMPLICIT NONE

   REAL(8), DIMENSION(:,:), ALLOCATABLE :: A, B, C, tC
   INTEGER :: L, M, N, N0, N1
   REAL ::   cput2, cput1

   INTEGER :: i, j, k, t, st
   REAL(8) :: rmin, rmax
   
!  OMP variables
   INTEGER :: OMP_GET_THREAD_NUM, OMP_GET_NUM_THREADS
   REAL(8) :: OMP_GET_WTIME, wall1, wall2
   INTEGER :: my_rank, numprocs
!
   ! Input Data
   WRITE(*,"(A)") " Input dimensions of matrices  L, M, N "
   READ*,L, M, N
   WRITE(*,"(A,3(1x,I6))") " Matrices dimensions are: ",L,M,N

   !  Master allocates entire matrices
   ALLOCATE(A(L,M), B(M,N), C(L,N), STAT=st)
   IF ( st /= 0 ) THEN
   	  WRITE(*,*) "Error allocating matrices "
   	  STOP
   END IF


   numprocs=1
   my_rank=0
   
   A = 0.0; B = 0.0; C = 0.0

   ! Every process executes
   !  Initializations
   
   ! Do it in parallel 
   DO i = 1, L
      DO k = 1, M
         A(i,k) = REAL(i-k) / REAL(i+k)
      END DO
   END DO

   ! Do it in parallel 
   DO j = 1, N
      DO k = 1, M
         B(k,j) = 0.01 * REAL(k-j) / REAL(k+j)
      END DO
   END DO

   !This statement is executed only by master thread if compiler openmp flag is used
   !$ wall1 = OMP_GET_WTIME()
   CALL CPU_TIME(cput1)
   
   ! Insert parallel region, pay attention to  shared or private variables

   !This statement is executed only by master thread if compiler openmp flag is used
   ! numprocs = OMP_GET_NUM_THREADS()
   ! my_rank = OMP_GET_THREAD_NUM()
   
   
   
   CALL intervals(N,N0,N1,numprocs,my_rank)
   !  Compute multiplication
   !This statement is executed by all threads but only one at a time
      print*," Process ",my_rank," of ",numprocs," computes C(:,",N0,",",N1,")"
   
   	   
   DO j = N0, N1
      DO i = 1, L
         C(i,j) = 0.0
         DO k = 1, M
            C(i,j) = C(i,j) + A(i,k) * B(k,j)
         END DO
      END DO
   END DO
   ! Close parallel region

   CALL CPU_TIME(cput2)
!   This statement is executed only by master thread if compiler openmp flag is used
   !wall2 = OMP_GET_WTIME()

   ! Check for correctness
   ALLOCATE(tC(L,N), STAT=st)
   IF ( st /= 0 ) THEN
      	 WRITE(*,*) "Error allocating matrix tC !!!"
      	 STOP
   END IF
      
      	   	   
      tC = MATMUL(A,B)

      rmin = MINVAL(tc-C); rmax = MAXVAL(tC-C)

      WRITE(*,*) " rmin, rmax = ",rmin,rmax
    
      WRITE(*,*) "Cpu and elapsed time are: ",(cput2-cput1),(wall2-wall1)
!
   STOP
END PROGRAM MatrMult  

   SUBROUTINE intervals(length,first,last,segments,n)
      IMPLICIT NONE
      INTEGER, INTENT(IN) :: length,segments,n
      INTEGER, INTENT(OUT) :: first,last

      INTEGER :: delta

      delta = ( length + (segments - 1) ) / segments
      first = delta * n + 1; last = delta * ( n + 1)
      IF ( last > length ) last = length
   
      RETURN
   END SUBROUTINE intervals

